### PR TITLE
Move tracking of walsenders out of Timeline.

### PR DESCRIPTION
**Pull request by arssher**
Refactors walsenders out of timeline.rs to makes it less convoluted into separate WalSenders with its own lock, but otherwise having the same structure. Tracking of in-memory remote_consistent_lsn is also moved there as it is mainly received from pageserver.

State of walsender (feedback) is also restructured to be cleaner; now it is either PageserverFeedback or StandbyFeedback(StandbyReply, HotStandbyFeedback), but not both.

Second commit moves PageserverFeedback to utils.
It allows to replace u64 with proper Lsn and pretty print PageserverFeedback
with serde(_json). Now walsenders on safekeepers queried with debug_dump look
like
```
"walsenders": [
  {
    "ttid": "fafe0cf39a99c608c872706149de9d2a/b4fb3be6f576935e7f0fcb84bdb909a1",
    "addr": "127.0.0.1:48774",
    "conn_id": 3,
    "appname": "pageserver",
    "feedback": {
      "Pageserver": {
	"current_timeline_size": 32096256,
	"last_received_lsn": "0/2415298",
	"disk_consistent_lsn": "0/1696628",
	"remote_consistent_lsn": "0/0",
	"replytime": "2023-04-12T13:54:53.958856+00:00"
      }
    }
  }
],
```

### CODE REVIEW

File: libs/utils/src/pageserver_feedback.rs [hunk 1]
```diff
@@ -0,0 +1,196 @@
...
+    //
+    // Following layout is used:
+    // char - number of key-value pairs that follow.
+    //
+    // key-value pairs:
+    // null-terminated string - key,
+    // uint32 - value length in bytes
+    // value itself
+    //
+    // TODO: change serialized fields names once all computes migrate to rename.
```
Comment (SomeoneToIgnore) [inline libs/utils/src/pageserver_feedback.rs:72]: We've changed the time format in the feedback, it's a breaking change and we can fix that TODO along the way then?

File: libs/utils/Cargo.toml [hunk 2]
```diff
@@ -35,6 +36,7 @@ strum_macros.workspace = true
 url.workspace = true
 uuid = { version = "1.2", features = ["v4", "serde"] }
 
+pq_proto.workspace = true
```
Comment (SomeoneToIgnore) [inline libs/utils/Cargo.toml:39]: So far, `pq_proto` was used in binary crates only and a `libs/postgres_backend` library crate.
`utils` is relatively bloated already, and there's been slow talks about breaking it up into pieces.

Should we put `pageserver_feedback.rs` into `postgres_backend` instead?

File: safekeeper/src/timeline.rs [hunk 3]
```diff
@@ -684,11 +581,21 @@ impl Timeline {
...
+        shared_state.get_safekeeper_info(
+            &self.ttid,
+            conf,
+            self.walsenders.get_remote_consistent_lsn(),
+        )
     }
 
     /// Update timeline state with peer safekeeper data.
-    pub async fn record_safekeeper_info(&self, sk_info: &SafekeeperTimelineInfo) -> Result<()> {
+    pub async fn record_safekeeper_info(&self, sk_info: &mut SafekeeperTimelineInfo) -> Result<()> {
```
Comment (SomeoneToIgnore) [inline safekeeper/src/timeline.rs:592]: Nothing added here needs `&mut` or am I wrong?

File: safekeeper/src/send_wal.rs [hunk 4]
```diff
@@ -51,24 +60,293 @@ impl HotStandbyFeedback {
...
+    /// and ts.
+    fn update_hs_feedback(&mut self) {
+        let mut agg = HotStandbyFeedback::empty();
+        for ws_state in self.slots.iter().flatten() {
+            if let ReplicationFeedback::Standby(standby_feedback) = ws_state.feedback {
+                let hs_feedback = standby_feedback.hs_feedback;
+                // doing Option math like op1.iter().chain(op2.iter()).min()
+                // would be nicer, but we serialize/deserialize this struct
+                // directly, so leave as is for now
+                if hs_feedback.xmin != INVALID_FULL_TRANSACTION_ID {
```
Comment (SomeoneToIgnore) [inline safekeeper/src/send_wal.rs:261]: IMO, there's plenty of room to unit test how the rules are applied, do you think we can add a couple for every feedback update? 🙂

**Review by SomeoneToIgnore — COMMENTED** [review]
Not sure I've understood the safekeeper part well, generally looks good, added a few notes.

File: libs/utils/src/pageserver_feedback.rs (hunk 1 above)
Comment (arssher) [reply libs/utils/src/pageserver_feedback.rs:72]: We need to ensure all old computes are gone for that. I'll probably look there (we don't have ready dashboard for that), but it is independent of this PR.

File: libs/utils/Cargo.toml (hunk 2 above)
Comment (arssher) [reply libs/utils/Cargo.toml:39]: Yeah, I'm also not a fan of that. But I'm not sure what would be better here. I tried moving pageserver_feedback.rs into postgres_backend, but the former depends on Lsn aka lsn.rs, and making pq_proto depend on utils sounds even worse. We could try separting 300 hundred lines Lsn into its own crate, but then lsn.rs depends on seqwait.rs. We could separate it as well, but is it worth it?

File: safekeeper/src/timeline.rs (hunk 3 above)
Comment (arssher) [reply safekeeper/src/timeline.rs:592]: No, its remote_consistent_lsn field is updated by taking max with local value.

Comment (petuhovskiy) [reply safekeeper/src/timeline.rs:592]: Maybe it's better to... 
```suggestion
    pub async fn record_safekeeper_info(&self, sk_info: mut SafekeeperTimelineInfo) -> Result<()> {
```

...to remove the reference to consume the value?

File: safekeeper/src/send_wal.rs (hunk 4 above)
Comment (arssher) [reply safekeeper/src/send_wal.rs:261]: Added some. Though I was a bit reluctant as we currently don't really do multiple pageserver / multiple standbys, and with their introduction policy can change.

File: libs/utils/Cargo.toml (hunk 2 above)
Comment (SomeoneToIgnore) [reply libs/utils/Cargo.toml:39]: Not sure: the API entities like these would be good to move away eventually, but it's not really the task of this PR indeed.

File: safekeeper/src/timeline.rs (hunk 3 above)
Comment (arssher) [reply safekeeper/src/timeline.rs:592]: changed

**Comment by arssher** [comment]
Test hopefully have been fixed, rebased. Should be ready for merge now.

**Review by petuhovskiy — APPROVED** [review]

**Pull request merged**